pkg/internal/codegen/parse/parser.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parse
+
+import (
+	"bufio"
+	"fmt"
+	"go/build"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/markbates/inflect"
+	"github.com/pkg/errors"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/gengo/args"
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen"
+)
+
+type APIs struct {
+	context         *generator.Context
+	arguments       *args.GeneratorArgs
+	Domain          string
+	VersionedPkgs   sets.String
+	UnversionedPkgs sets.String
+	APIsPkg         string
+	APIsPkgRaw      *types.Package
+	GroupNames      sets.String
+
+	APIs        *codegen.APIs
+	Controllers []codegen.Controller
+
+	ByGroupKindVersion    map[string]map[string]map[string]*codegen.APIResource
+	ByGroupVersionKind    map[string]map[string]map[string]*codegen.APIResource
+	SubByGroupVersionKind map[string]map[string]map[string]*types.Type
+	Groups                map[string]types.Package
+	Rules                 []rbacv1.PolicyRule
+	Informers             map[v1.GroupVersionKind]bool
+}
+
+func NewAPIs(context *generator.Context, arguments *args.GeneratorArgs, domain string) *APIs {
+	b := &APIs{
+		context:   context,
+		arguments: arguments,
+		Domain:    domain,
+	}
+	b.parsePackages()
+	b.parseGroupNames()
+	b.parseIndex()
+	b.parseControllers()
+	b.parseRBAC()
+	b.parseInformers()
+	b.verifyRBACAnnotations()
+	b.parseAPIs()
+	b.parseCRDs()
```

Comment:
We need to revisit this. Lets only do the common bits of parsing here. RBAC/CRD should be moved out.

Reply:
I would expect the parser to produce CRD objects for command to use.
If we remove it here, then where do you think is the proper place to do the transformation?

Reply:
This is fine for now. Lets get this is in `as-is` and remove the unnecessary pieces in a separate PR.